sources/web/datalab/auth.ts

```diff
@@ -0,0 +1,153 @@
+ */
+function persistCredentials(tokens: any) {
+  if (!fs.existsSync(gcloudDir)) {
+    fs.mkdirSync(gcloudDir);
+  }
+  fs.writeFileSync(appCredFile,
+      JSON.stringify({
+        client_id: clientId,
+        client_secret: clientSecret,
+        refresh_token: tokens.refresh_token,
+        type: "authorized_user"
+      })
+  );
+  var segments = tokens.id_token.split('.');
+  var payload = JSON.parse(base64decodeSegment(segments[1]));
+  fs.writeFileSync(userCredFile,
+      JSON.stringify({
+        data:
+            [
+              {
+                credential: {
+                  _class: "OAuth2Credentials",
+                  _module: "oauth2client.client",
+                  access_token: tokens.access_token,
+                  client_id: clientId,
+                  client_secret: clientSecret,
+                  id_token: payload,
+                  invalid: false,
+                  refresh_token: tokens.refresh_token,
+                  revoke_uri: "https://accounts.google.com/o/oauth2/revoke",
+                  token_expiry: (new Date(tokens.expiry_date)).toISOString(),
+                  token_response: {
+                    access_token: tokens.access_token,
+                    expires_in: 3600,
+                    id_token: tokens.id_token,
+                    refresh_token: tokens.refresh_token,
+                    token_type: "Bearer"
+                  },
+                  token_uri: "https://accounts.google.com/o/oauth2/token",
+                  user_agent: "Cloud SDK Command Line Tool"
+                },
+                key: {
+                  account: payload.email,
+                  type: "google-cloud-sdk"
+                }
+              }
+            ],
+        file_version: 1
+      })
+  );
+  // Tell gcloud which account we are using.
+  childProcess.exec('gcloud config set account ' + payload.email, {env: process.env}, function(err, stdout, stderr) {
+    if (err) {
+      logging.getLogger().error(err, 'Failed to set gcloud account. stderr: %s', stderr);
+      return;
+    }
+  });
+}
+
+export function handleAuthFlow(request: http.ServerRequest, response: http.ServerResponse, parsed_url: any): boolean {
+  var path = parsed_url.pathname;
+  if (path.indexOf('/oauthcallback') == 0) {  // Return from auth flow.
+    var query = parsed_url.query;
+    if (query.code) {
+      oauth2Client.getToken(query.code, function (err:any, tokens:any) {
+        if (err) {
+          response.writeHead(403);
+          response.end();
+        } else {
+          logging.getLogger().info('Got tokens');
+          oauth2Client.setCredentials(tokens);
+          // Push them to Jupyter and handle request.
+          persistCredentials(tokens);
+          response.statusCode = 302;
+          response.setHeader('Location', query.state);
+          response.end();
+        }
+      });
+    }
+  } else if (!oauth2Client && process.env.DATALAB_ENV == 'local' && !fs.existsSync(userCredFile)) {
```

Comment:
I assume on GCE auth flow will be skipped because DATALAB_ENV is not set to 'local'?